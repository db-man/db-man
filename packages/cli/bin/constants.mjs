export const DB_CFG_FILENAME = 'dbcfg.json';
export const ERR_NOT_FOUND_TABLE = 1;
export const ERR_NO_PRIMARY_KEY = 2;

// Configurable table file suffix
export const TABLE_DATA_FILE_SUFFIX = '.data.json';
