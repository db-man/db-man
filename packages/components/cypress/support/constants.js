// CYPRESS_DBM_GH_TOKEN=123 npx cypress open
export const ghToken = Cypress.env('DBM_GH_TOKEN');

export const ghConfig = {
  token: Cypress.env('DBM_GH_TOKEN'),
  repoOwner: 'db-man',
  repoName: 'split-table-db',
  repoPath: 'db_files_dir',
  dbModes: 'split-table',
};
