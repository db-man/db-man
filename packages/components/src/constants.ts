import { types } from '@db-man/github';

// types
export const STRING_ARRAY: types.DbColumnType = 'STRING_ARRAY';
export const STRING: types.DbColumnType = 'STRING';
export const NUMBER = 'NUMBER';
export const BOOL = 'BOOL';

// Localstorage keys
// DB info
export const LS_KEY_DB_CONNECTIONS = 'dbm_db_connections';
export const LS_KEY_GITHUB_PERSONAL_ACCESS_TOKEN =
  'dbm_github_personal_access_token';
export const LS_KEY_GITHUB_OWNER = 'dbm_github_owner'; // owner: db-man
export const LS_KEY_GITHUB_REPO_NAME = 'dbm_github_repo_name'; // repo: split-table-db
export const LS_KEY_GITHUB_REPO_PATH = 'dbm_github_repo_path'; // repoPath: db_files_dir
export const LS_KEY_GITHUB_REPO_MODES = 'dbm_github_repo_modes'; // dbModes: split-table
export const LS_KEY_DBS_SCHEMA = 'dbm_dbs_schema';
// other switches
export const LS_IS_DARK_THEME = 'dbm_is_dark_theme';
export const LS_SHOW_DOWNLOAD_BUTTON = 'dbm_show_download_button';
// UI state
export const LS_QUERY_PAGE_SELECTED_TABLE_NAMES =
  'dbm_query_page_selected_table_names';

// Table column definitions
export const COL_UI_PRESETS = 'ui:createUpdatePage:presets';
export const COL_UI_LISTPAGE_RANDOMVIEW = 'ui:listPage:randomView';

// UI type
export const TYPE_CREATE_UPDATE_PAGE = 'type:createUpdatePage';
export const TYPE_GET_PAGE = 'type:getPage';
export const TYPE_LIST_PAGE = 'type:listPage';

export const DBS_CFG_FILENAME = 'dbs.json';
export const DB_CFG_FILENAME = 'dbcfg.json';
