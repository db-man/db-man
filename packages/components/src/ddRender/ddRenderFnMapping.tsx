/* eslint-disable max-len */

import React from 'react';
/**
 * When importing like this `import Handlebars from 'handlebars'`, will see below error when running npm run dev
 *
 * ```
 * Failed to compile.
 *
 * Module not found: Error: Can't resolve 'fs' in '/Users/devin.chenyang/source/github.com/db-man/components/node_modules/handlebars/lib'
 * WARNING in ./node_modules/handlebars/lib/index.js 20:38-56
 * require.extensions is not supported by webpack. Use a loader instead.
 *
 * WARNING in ./node_modules/handlebars/lib/index.js 21:2-20
 * require.extensions is not supported by webpack. Use a loader instead.
 *
 * WARNING in ./node_modules/handlebars/lib/index.js 22:2-20
 * require.extensions is not supported by webpack. Use a loader instead.
 * ```
 *
 * The current tmp solution is to use `import Handlebars from 'handlebars/dist/handlebars'` instead.
 *
 * See more: https://github.com/handlebars-lang/handlebars.js/issues/1174
 */
// import Handlebars from 'handlebars';
// @ts-ignore Ignore handlebars type error
import Handlebars from 'handlebars/dist/handlebars';

import {
  ImageLink,
  ImageLinks,
  Link,
  Links,
  Fragment,
} from '../components/Links';
import PhotoList from '../components/PhotoList';
import ErrorAlert from '../components/ErrorAlert';
import TextAreaFormFieldValue from '../components/TextAreaFormFieldValue';
import { RowType } from '../types/Data';
import DbTable from '../types/DbTable';
import { RenderArgs } from '../types/UiType';
import { getTablePrimaryKey } from '../dbs';

interface Options {
  fn: (this: any) => string;
  hash: {
    rows?: RowType[];
    tables: DbTable[];
    tableName: string;
    primaryKeyVal: any;
  };
}

/**
 * tpl: {{#replace "foo" "bar"}}{{title}}{{/replace}}
 * input: {title:"foo"}
 * output: bar
 */
Handlebars.registerHelper(
  'replace',
  function replaceHelper(find: string, replace: string, options: Options) {
    // @ts-ignore
    const string = options.fn(this);
    return string.replace(find, replace);
  }
);

/**
 * tpl: {{join record.tags ", "}}
 * input: {"tags": ["foo", "bar"]}
 * output: "foo, bar"
 */
Handlebars.registerHelper('join', (arr: string[], sep: string) => {
  if (!arr) return '';
  return arr.join(sep);
});

Handlebars.registerHelper('getTableRecordByKey', (options: Options) => {
  if (!options.hash.rows) return null;
  const primaryKey = getTablePrimaryKey(
    options.hash.tables,
    options.hash.tableName
  );
  return options.hash.rows.find(
    (row: RowType) => row[primaryKey] === options.hash.primaryKeyVal
  );
});

/**
 * Generate a render function based on the passing component,
 * so we can later call this render function by passing an Handlebars template to dynamic render component based on the passing record
 * @param {*} Component
 * @returns {()=><Component/>} render function
 */
const genRenderFunc = (Component: any) =>
  function dComponent(
    val: any,
    record: RowType,
    index?: number,
    args?: RenderArgs,
    tplExtra?: any
  ) {
    if (!record || typeof record !== 'object') {
      console.error('[genRenderFunc] record should be an object!', record); // eslint-disable-line no-console
    }
    if (!args || !args[1]) {
      return <Component>{val}</Component>;
    }

    const tplStr = args[1];
    const tpl = Handlebars.compile(tplStr);
    const json = tpl({ record, extra: tplExtra });
    try {
      // {foo:'bar'} <= "{\"foo\":\"bar\"}"
      // "foo" <= "\"foo\""
      const props = JSON.parse(json);
      if (typeof props === 'string') {
        return <Component>{props}</Component>;
      }
      return <Component {...props} />; // eslint-disable-line react/jsx-props-no-spreading
    } catch (err) {
      console.error('Failed to parse JSON for tpl, err:', err, json);
      return (
        <div>
          val: {val}
          <ErrorAlert
            json={json}
            error={err as Error}
            tplStr={tplStr}
            record={record}
          />
        </div>
      );
    }
  };

type DdRenderFnMappingType = Record<
  string, // render function name, e.g. "Link"
  (
    val: any,
    record: RowType,
    index?: number,
    args?: RenderArgs,
    tplExtra?: any
  ) => JSX.Element
>;

/**
 * Data Driving Render Function Mapping
 */
const ddRenderFnMapping: DdRenderFnMappingType = {
  // renderFnName: (cellVal, rowVal, colIndex, ...renderFnStr) => null
  // renderFnName: (cellVal, rowVal, colIndex, ...renderFnProps) => null

  /**
   * @param {number|string} val e.g. "docs"
   * @param {Object} record e.g. `{"org":"github","repo":"docs"}`
   * @param {number} index e.g. 9, it's the 10th row
   * @param {Object|undefined} args e.g `["Link", "{\"href\":\"https://github.com/{{record.org}}/{{record.repo}}\",\"text\":\"{{record.repo}}\"}"]`
   *                                args could be undefined when column is {"type:listPage": "Link"}
   *  Usage:
   * 1. "type:getPage": ["Link", "{\"href\":\"https://github.com/{{record.org}}/{{record.repo}}\",\"text\":\"{{record.repo}}\"}"],
   * 2. "type:listPage": ["Link", "{\"href\":\"https://github.com/{{record.org}}/{{record.repo}}\",\"text\":\"{{record.repo}}\"}"]
   * 3. "type:listPage": "Link"
   */
  Link: genRenderFunc(Link),
  /**
   * Usage:
   * ```json
   * {"type:listPage": ["ImageLink",
   *   "{\"url\":\"{{record.url}}\",\"imgSrc\":\"{{record.url}}_th.jpg\",
   *     \"tags\":\"{{record.tags}}\"}"
   * ]}
   * ```
   * Issues:
   *   When record.url="https://foo.com/get?foo=bar".
   *   By using `{{record.url}}` will generate "https://foo.com/get?foo&#x3D;bar".
   *   To prevent this issue, `{{{record.url}}}` could be used. (https://handlebarsjs.com/guide/expressions.html#html-escaping)
   */
  ImageLink: genRenderFunc(ImageLink),
  ImageLinks: genRenderFunc(ImageLinks),
  /**
   * Usage:
   * ```json
   * {"type:listPage": ["Fragment", "{{join record.tags \", \"}}"]}
   * ```
   */
  Fragment: genRenderFunc(Fragment),
  Links: genRenderFunc(Links),
  /**
   * Usage:
   * ```json
   * {
   *   "type:getPage": [
   *     "PhotoList",
   *     "[{{#each record.photoUrls}}{{#if @index}},{{/if}}{\"url\":\"{{this}}\",\"imgSrc\":\"{{this}}_th.jpg\",\"description\":\"{{#with (getTableRecordByKey tables=../extra.tables tableName=\"imgs\" primaryKeyVal=this rows=../extra.rows)}}{{join tags \", \"}}{{/with}}\"}{{/each}}]"
   *   ]
   * }
   * ```
   */
  PhotoList: genRenderFunc(PhotoList),
  TextAreaFormFieldValue: genRenderFunc(TextAreaFormFieldValue),
};

export default ddRenderFnMapping;
