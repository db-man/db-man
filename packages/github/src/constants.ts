export const DBS_CFG_FILENAME = 'dbs.json';
export const DB_CFG_FILENAME = 'dbcfg.json';
