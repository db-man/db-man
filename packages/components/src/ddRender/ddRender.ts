import * as constants from '../constants';
import DbColumn, { RenderKeyType } from '../types/DbColumn';
import { RowType } from '../types/Data';
import { RenderArgs } from '../types/UiType';
import ddRenderFnMapping from './ddRenderFnMapping';

// Default render func when "type:listPage" or "type:getPage" not defined in db table column
const defaultRendersForListPage = {
  [constants.NUMBER]: (val: number) => val,
  [constants.STRING]: (val: string) => val,
  [constants.STRING_ARRAY]: (val: string[]) => val && val.join(', '),
  [constants.BOOL]: (val: boolean) => (val === undefined ? '' : String(val)),
};
const defaultRendersForGetPage = {
  [constants.NUMBER]: (val: number) =>
    val === undefined ? 'NO_VALUE' : String(val),
  [constants.STRING]: (val: string) => (val === undefined ? 'NO_VALUE' : val),
  [constants.STRING_ARRAY]: (val: string[]) =>
    val === undefined ? 'NO_VALUE' : val && val.join(', '),
  [constants.BOOL]: (val: boolean) =>
    val === undefined ? 'NO_VALUE' : String(val),
};

/**
 * @param {string|string[]} args e.g. "Link" or ["Link", "{{record.url}}"]
 * @param {Object} tplExtra
 */
export const getRender = (args: RenderArgs, tplExtra?: any) => {
  // the column render function defined in Table component of antd
  // renderFn = (val, record, index) => ()
  let renderFn;

  if (!args) {
    return renderFn;
  }

  if (typeof args === 'string') {
    const fn = ddRenderFnMapping[args];
    if (fn) {
      renderFn = fn;
    }
  }

  if (Array.isArray(args)) {
    const [renderFnName] = args;
    renderFn = (val: any, record: RowType, index?: number) =>
      ddRenderFnMapping[renderFnName](val, record, index, args, tplExtra);
  }

  return renderFn;
};

type ColumnRenderType = (
  renderKey: RenderKeyType,
  column: DbColumn,
  tplExtra?: any
) => (val: any, record: RowType, index?: number) => any;

/**
 * column def:
 * {
 *   "type:listPage": ["Link", "{{record.url}}"]
 * }
 */
export const getColumnRender: ColumnRenderType = (
  renderKey,
  column,
  tplExtra
) => {
  // should only used for "type:listPage" or "type:getPage"
  if (
    !renderKey ||
    [constants.TYPE_LIST_PAGE, constants.TYPE_GET_PAGE].indexOf(renderKey) < 0
  ) {
    console.error('getColumnRender: invalid renderKey', renderKey);
  }

  const customRender = getRender(column[renderKey], tplExtra);
  if (customRender) {
    return customRender;
  }

  if (renderKey === constants.TYPE_GET_PAGE) {
    return defaultRendersForGetPage[column.type || constants.STRING];
  }

  // the column render function defined in Table component of antd
  // renderFn = (val, record, index) => ()
  return defaultRendersForListPage[column.type || constants.STRING];
};

// export const getRenderResultByColumn = (
//   value: any,
//   record: RowType,
//   index: number,
//   args: RenderArgs,
//   column: Column
// ) => getColumnRender(constants.TYPE_LIST_PAGE, column)(value, record, index);
