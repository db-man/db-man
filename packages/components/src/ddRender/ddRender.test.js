import Handlebars from 'handlebars/dist/handlebars';
import { getRender, getColumnRender } from './ddRender';
import { TYPE_GET_PAGE } from '../constants';

describe('getColumnRender', () => {
  test('should return proper value', () => {
    const view = getColumnRender('type:listPage', {
      id: 'url',
      'type:listPage': 'ImageLink',
    });
    expect(
      view('https://a.com/b.jpg', { url: 'https://a.com/b.jpg' }, 1)
    ).toMatchSnapshot();
  });

  test('should return proper value2', () => {
    const view = getColumnRender('type:listPage', {
      id: 'yrl',
      'type:listPage': [
        'ImageLink',
        '{"url":"{{record.url}}","imgSrc":"{{record.url}}"}',
      ],
    });
    expect(
      view('https://a.com/large/b.jpg', { url: 'https://a.com/large/b.jpg' })
    ).toMatchSnapshot();
  });

  describe('given Link', () => {
    it('should match snapshot', () => {
      expect(
        getColumnRender('type:listPage', {
          'type:listPage': [
            'Link',
            '{"href":"{{record.url}}","text":"{{record.url}}"}',
          ],
        })('https://foo.com', { url: 'https://foo.com' })
      ).toMatchSnapshot();
    });
  });

  describe('given ImageLink', () => {
    it('should match snapshot', () => {
      expect(
        getColumnRender('type:listPage', {
          'type:listPage': [
            'ImageLink',
            '{"url":"{{record.photos.[0]}}","imgSrc":"{{record.photos.[0]}}"}',
          ],
        })('https://foo.com/a.jpg', {
          id: 'foo',
          photos: ['https://foo.com/a.jpg'],
        })
      ).toMatchSnapshot();
    });
    it('should match snapshot when photos is undefined', () => {
      expect(
        getColumnRender('type:listPage', {
          'type:listPage': [
            'ImageLink',
            '{"url":"{{record.photos.[0]}}","imgSrc":"record.photos.[0]"}',
          ],
        })('', { id: 'foo' })
      ).toMatchSnapshot();
    });
  });

  describe('given hidden column in a list page', () => {
    it('should return default render func', () => {
      const view = getColumnRender('type:listPage', {
        'type:listPage': 'HIDE',
      });
      expect(view('foo')).toBe('foo');
    });
  });

  describe('default render function', () => {
    it('type is STRING', () => {
      const view = getColumnRender(TYPE_GET_PAGE, {
        id: 'name',
        name: 'Name',
        type: 'STRING',
      });
      expect(view('David', { name: 'David' })).toBe('David');
      expect(view(undefined, { name: undefined })).toBe('NO_VALUE');
    });
    it('type is STRING_ARRAY', () => {
      const view = getColumnRender(TYPE_GET_PAGE, {
        id: 'tags',
        name: 'Tags',
        type: 'STRING_ARRAY',
      });
      expect(view(['dog', 'cat'], { tags: ['dog', 'cat'] })).toBe('dog, cat');
      expect(view(undefined, { tags: undefined })).toBe('NO_VALUE');
    });
    it('type is NUMBER', () => {
      const view = getColumnRender(TYPE_GET_PAGE, {
        id: 'age',
        name: 'Age',
        type: 'NUMBER',
      });
      expect(view(12, { age: 12 })).toBe('12');
      expect(view(undefined, { age: undefined })).toBe('NO_VALUE');
    });
    it('type is BOOL', () => {
      const view = getColumnRender(TYPE_GET_PAGE, {
        id: 'active',
        name: 'Active',
        type: 'BOOL',
      });
      expect(view(true, { active: true })).toBe('true');
      expect(view(undefined, { active: undefined })).toBe('NO_VALUE');
    });
  });
});

describe('getRender', () => {
  describe('given simple arg', () => {
    it('should render properly', () => {
      const args = 'Link';
      const record = { url: 'https://foo.com' };
      expect(getRender(args)(record.url, record)).toMatchSnapshot();
    });
  });
  describe('given tpl', () => {
    it('should render properly', () => {
      const args = [
        'Link',
        '{"href":"{{record.url}}","text":"{{record.url}}"}',
      ];
      const record = { url: 'https://foo.com' };
      expect(getRender(args)(record.url, record)).toMatchSnapshot();
    });
  });
  describe('TextAreaFormFieldValue', () => {
    it('should match snapshot', () => {
      const args = [
        'TextAreaFormFieldValue',
        '{"label":"{{extra.column.name}}","rows":2,"value":"{{record.note}}"}',
      ];
      const column = { id: 'note', name: 'Note' };
      const record = { note: 'This is TODO' };
      expect(
        getRender(args, { column })(record.note, record)
      ).toMatchSnapshot();
    });
  });
});

describe('handlebars helper getTableRecordByKey', () => {
  it('should return proper value', () => {
    const tables = [
      { name: 'users', columns: [{ id: 'userId', primary: true }] },
      { name: 'imgs', columns: [{ id: 'url', primary: true }] },
    ];
    const imgsTableRows = [
      { url: 'https://foo.jpg', tags: ['foo', 'bar'] },
      { url: 'https://bar.jpg', tags: ['foo2', 'bar2'] },
    ];
    const userRecord = {
      userId: '123',
      photos: ['https://foo.jpg', 'https://bar.jpg'],
    };
    // loop each photos of one user from `users` table
    // for each photo link, try to search in `imgs` table
    // when found, get the tags of record from `imgs` table
    const tpl = Handlebars.compile(`[
{{#each userRecord.photos}}
{{#if @index}},{{/if}}
{
  "url":"{{this}}",
  "imgSrc":"{{this}}_th.jpg",
  "description":"{{#with (
    getTableRecordByKey
    tables=../extra.tables
    tableName="imgs"
    primaryKeyVal=this
    rows=../extra.rows
  )}}{{join tags ", "}}{{/with}}"
}
{{/each}}
]`);
    const json = tpl({
      userRecord,
      extra: {
        tables,
        rows: imgsTableRows,
      },
    });
    const result = JSON.parse(json);
    expect(result).toHaveLength(2);
    expect(result[0].url).toBe('https://foo.jpg');
    expect(result[0].description).toBe('foo, bar');
  });
});
