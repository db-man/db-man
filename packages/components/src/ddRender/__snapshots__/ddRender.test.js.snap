// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getColumnRender given ImageLink should match snapshot 1`] = `
<ImageLink
  imgSrc="https://foo.com/a.jpg"
  url="https://foo.com/a.jpg"
/>
`;

exports[`getColumnRender given ImageLink should match snapshot when photos is undefined 1`] = `
<ImageLink
  imgSrc="record.photos.[0]"
  url=""
/>
`;

exports[`getColumnRender given Link should match snapshot 1`] = `
<Link
  href="https://foo.com"
  text="https://foo.com"
/>
`;

exports[`getColumnRender should return proper value 1`] = `
<ImageLink>
  https://a.com/b.jpg
</ImageLink>
`;

exports[`getColumnRender should return proper value2 1`] = `
<ImageLink
  imgSrc="https://a.com/large/b.jpg"
  url="https://a.com/large/b.jpg"
/>
`;

exports[`getRender TextAreaFormFieldValue should match snapshot 1`] = `
<TextAreaFormFieldValue
  label="Note"
  rows={2}
  value="This is TODO"
/>
`;

exports[`getRender given simple arg should render properly 1`] = `
<Link>
  https://foo.com
</Link>
`;

exports[`getRender given tpl should render properly 1`] = `
<Link
  href="https://foo.com"
  text="https://foo.com"
/>
`;
